framework/math/ParallelVector/ParallelSTLVector.cc

```diff
@@ -0,0 +1,479 @@
+#include "ParallelSTLVector.h"
```

Comment:
There are a few different reasons I originally implemented this. For some background, in @Naktakala implementation of the source corrected DSA k-eigenvalue solver, he put in an option to do PWLC diffusion. This required access to PWLD ghost DoFs. To do that, he created a ghost communicator that was stored within the solver, and then when he needed the PWLD ghosts, he created a local vector, and then used the communicator to populate the ghosts at the back and communicate. I needed similar functionality for a second moment method solver that will be put in the code base in the future. Here two reasons for this:

1. The GhostedParallelSTLVector encapsulated the vector ghost communicator and the actual data in one structure which made the task of getting PWLD ghosts a little more streamlined (and makes the code a bit easier to read). In addition to the PWLD -> PWLC mapping, this was needed for accessing nonlocal tensors for the second moment method.
2. The ParallelSTLVector was a natural thing to implement as a parent of the ghosted version, but how this is currently used is to construct a source vector for a diffusion solver outside of the solver that needs to contribute to nonlocal entries. This source then gets added into the diffusion solver PETSc source vector with the existing routines for adding an STL vector to the diffusion source vector.
3. This could be a platform to have a common interface for vector types from a variety of different packages.

We could certainly use PETSc vectors throughout the entire library if we wanted, so these implementations do nothing other than provide a nicety for a user. It also served as a good learning opportunity for me to get my feet wet in the weeds of this stuff.

Reply:
@zachhardy I do not fully understand this, especially in light of your last paragraph. My opinion is that we should let PETSc do what it is good at (and its team of developers). Could this have bee a nicer user interface that simply wraps around existing PETSc functionalities? Are we just bloating the base code?

Reply:
I agree with @ragusa.  This is my biggest concern.  I could understand and get behind something along the lines of (3), but I'm concerned we will end up with something that requires a great deal of effort to maintain/develop and that just replicates functionality we can get from other packages like PETSC.

Reply:
@ragusa The GhostedParallelSTLVector improved upon existing code in ChiTech for ghosted STL vectors. The ParallelSTLVector was essentially given for free in implementing the prior.

I understand the concern here, but what is the logical end of that argument? You could use PETSc for almost everything in the code base, yet we use STL vectors and transfer to and from PETSc vectors when we are about to use a PETSc solver where we could just use PETSc vectors in the first place.

Reply:
Oh is this an actual serious question?? Have any of you actually used a FEM library where they didn't do this. MOOSE has it, deal.ii has it....

Any c++ developer that's been using PETSc for a while will complain about the following:

PETSc can get quite hectic! 
- It has open/close notions on vectors. 
- The `Vec` and `Mat` are actually typedef on pointers so ownership is never clear. You constantly have to worry about where you can call VecDestroy and MatDestroy
- Cannot use stuff like unique_ptr on them.
- Reading from them is always like 3 lines.
- The headers are absolutely gigantic. Just to get a .cc function to complete the type for Vec you need the petsc.h header, then that absolutely blows up the compile time.
- Ever tried their block vectors? Nightmare.

Geez yeah. No the abstraction in `ParallelVector` and the PETSc-free `ParallelSTLVector` is super valuable. It will benefit our successors 100%. I would waaaay rather transform a PETSc vector straight out of their solvers (to my own), then do my operations on a more c++-fashion vector, until I am ready for PETSc to do its thing, then hand it back. This way PETSc still does the heavy lifting. Only dip into the chaos (efficient chaos yes) momentarily and maintain more sane code outside of PETSc.

Reply:
Wow. Bot from X?

Reply:
?

Reply:
Bot from X????

Reply:
I have actually used an FEM library... every day in fact. But not PETSC. I'm not a big fan, and this just reinforces that. This seems like a tremendous amount of effort to work around PETSC's shortcomings.